packages/gatsby/src/utils/dev-ssr/render-dev-html.ts

```diff
@@ -0,0 +1,157 @@
+import JestWorker from "jest-worker"
+import fs from "fs-extra"
+import { joinPath } from "gatsby-core-utils"
+import report from "gatsby-cli/lib/reporter"
+import _ from "lodash"
+
+import { startListener } from "../../bootstrap/requires-writer"
+import { findPageByPath } from "../find-page-by-path"
+import { boundActionCreators } from "../../redux/actions"
+const { createSSRVisitedPage } = boundActionCreators
+
+const startWorker = (): any => {
+  const newWorker = new JestWorker(require.resolve(`./render-dev-html-child`), {
+    exposedMethods: [`renderHTML`, `deleteModuleCache`, `warmup`],
+    numWorkers: 1,
+    forkOptions: { silent: false },
+  })
+
+  // jest-worker is lazy with forking but we want to fork immediately so the user
+  // doesn't have to wait.
+  // @ts-ignore
+  newWorker.warmup()
+
+  return newWorker
+}
+
+let worker
+export const initDevWorkerPool = (): void => {
+  worker = startWorker()
+}
+
+let changeCount = 0
+export const restartWorker = (htmlComponentRendererPath): void => {
+  changeCount += 1
+  // Forking is expensive — each time we re-require the outputted webpack
+  // file, memory grows ~10 mb — 25 regenerations means ~250mb which seems
+  // like an accepatable amount of memory to grow before we reclaim it
+  // by rebooting the worker process.
+  if (changeCount > 25) {
+    const oldWorker = worker
+    const newWorker = startWorker()
+    worker = newWorker
+    oldWorker.end()
+    changeCount = 0
+  } else {
+    worker.deleteModuleCache(htmlComponentRendererPath)
+  }
+}
+
+const searchFileForString = (substring, filePath): Promise<boolean> =>
+  new Promise(resolve => {
+    const stream = fs.createReadStream(filePath)
+    let found = false
+    stream.on(`data`, function (d) {
+      if (d.includes(substring)) {
+        found = true
+        stream.close()
+        resolve(found)
+      }
+    })
+    stream.on(`error`, function () {
+      resolve(found)
+    })
+    stream.on(`close`, function () {
+      resolve(found)
+    })
+  })
+
+const ensurePathComponentInSSRBundle = async (
```

Comment:
This approach is a bit weird. I wasn't able to think of a better way however w/o deeper integration w/ webpack (I assume the webpack stats object would tell us what modules are in the bundle). When we work on this for develop-js — we should refactor everything into the state machine.

Reply:
Just checking every time though is fast (most checks took < 100ms) and reliable.